presto-orc/src/main/java/com/facebook/presto/orc/reader/MapFlatSelectiveStreamReader.java

```diff
@@ -0,0 +1,830 @@
+        this.outputRequired = requireNonNull(outputType, "outputType is null").isPresent();
+        this.outputType = outputType.map(MapType.class::cast).orElse(null);
+
+        TupleDomainFilter filter = getTopLevelFilter(filters).orElse(null);
+        this.nullsAllowed = filter == null || filter.testNull();
+        this.nonNullsAllowed = filter == null || filter.testNonNull();
+
+        requireNonNull(requiredSubfields, "requiredSubfields is null");
+        switch (keyOrcType) {
+            case BYTE:
+            case SHORT:
+            case INT:
+            case LONG:
+                requiredLongKeys = requiredSubfields.stream()
+                        .map(Subfield::getPath)
+                        .map(path -> path.get(0))
+                        .map(Subfield.LongSubscript.class::cast)
+                        .map(Subfield.LongSubscript::getIndex)
+                        .collect(toImmutableSet());
+                requiredStringKeys = null;
+                return;
+            case STRING:
+            case BINARY:
+                requiredStringKeys = requiredSubfields.stream()
+                        .map(Subfield::getPath)
+                        .map(path -> path.get(0))
+                        .map(Subfield.StringSubscript.class::cast)
+                        .map(Subfield.StringSubscript::getIndex)
+                        .collect(toImmutableSet());
+                requiredLongKeys = null;
+                return;
+            default:
+                throw new IllegalArgumentException("Unsupported flat map key type: " + keyOrcType);
+        }
+    }
+
+    private static Optional<TupleDomainFilter> getTopLevelFilter(Map<Subfield, TupleDomainFilter> filters)
+    {
+        Map<Subfield, TupleDomainFilter> topLevelFilters = Maps.filterEntries(filters, entry -> entry.getKey().getPath().isEmpty());
+        if (topLevelFilters.isEmpty()) {
+            return Optional.empty();
+        }
+
+        checkArgument(topLevelFilters.size() == 1, "MAP column may have at most one top-level range filter");
+        TupleDomainFilter filter = Iterables.getOnlyElement(topLevelFilters.values());
+        checkArgument(filter == IS_NULL || filter == IS_NOT_NULL, "Top-level range filter on MAP column must be IS NULL or IS NOT NULL");
+        return Optional.of(filter);
+    }
+
+    @Override
+    public int read(int offset, int[] positions, int positionCount)
+            throws IOException
+    {
+        checkState(!valuesInUse, "BlockLease hasn't been closed yet");
+
+        if (!rowGroupOpen) {
+            openRowGroup();
+        }
+
+        allNulls = false;
+
+        if (outputRequired && nullsAllowed && presentStream != null) {
+            nulls = ensureCapacity(nulls, positionCount);
+        }
+
+        if (!nullsAllowed || !nonNullsAllowed) {
+            outputPositions = ensureCapacity(outputPositions, positionCount);
```

Comment:
This is saying: If we are going to output any data, make sure we have enough capacity. If we aren't, then we are just filtering and not actually reading any values. 

Is that right? I'm a little confused by the business logic here

Reply:
This is saying: `if there is a filter, allocate memory for outputPositions; otherwise, use input positions as outputPositions`